Add doc for subscribed events and events workflows (#241)
vmware/precaution: docs/events_handling.md

# Events handling

## Subscribed events

Precaution needs to subscribe to a number of events
to satisfy different use cases.

> The events are separated into two groups `PullRequestEvent` and `CheckRunEvent/CheckSuiteEvent`.

Reviewer: ```suggestion
The events are separated into two groups `PullRequestEvent` and `CheckRunEvent and CheckSuiteEvent`.
```

### PullRequestEvent

> Official GitHub documentation: [PullRequest event](https://developer.github.com/v3/activity/events/types/#pullrequestevent)

Reviewer: Is it "PullRequest event" or "PullRequestEvent"?

Author: Yep, my mistake.

#### `pull_request.opened`

- __occurs__: when a new pull request is created

- __triggers__: start of the Precaution scan

#### `pull_request.reopened`

- __occurs__: when a pull request is reopened from a closed state

- __triggers__: start of a new Precaution scan

#### `pull_request.synchronize`

- __occurs__: when you add a commit to already opened pull request

- __triggers__: start of a new Precaution scan

> For more information look at pull request number [#113](https://github.com/vmware/precaution/pull/113)

Reviewer: ```suggestion
For more information look at pull request [Fix for "across the forks" PRs problems](https://github.com/vmware/precaution/pull/113)
```

### CheckRunEvent and CheckSuiteEvent

Official GitHub documentation:
- [CheckRunEvent](https://developer.github.com/v3/activity/events/types/#checkrunevent)
- [CheckSuiteEvent](https://developer.github.com/v3/activity/events/types/#checksuiteevent)

A user had already opened a pull request and is at the pull request page.

> He navigates to the "Checks" tab.

Reviewer: ```suggestion
They navigate to the "Checks" tab.
```

Reviewer: Still shows "he".  Is the user always a "he"?

> They navigates to the "Checks" tab.

Reviewer: ```suggestion
They navigate to the "Checks" tab.
```

> If the Precaution check had failed he will see "Re-run all checks", "Re-run" and

Reviewer: ```suggestion
If the Precaution check failed he will see "Re-run all checks", "Re-run" and
```

> If the Precaution check failed he will see "Re-run all checks", "Re-run" and

Reviewer: ```suggestion
If the Precaution check failed they will see "Re-run all checks", "Re-run" and
```

Reviewer: Still shows "he".  Is the user always a "he"?

"Re-run failed checks" buttons.

#### `check_run.rerequested`

- __occurs__: when you click the "Re-run" button

- __triggers__: start of a new Precaution scan

#### `check_suite.rerequested`

- __occurs__: when you click "Re-run all checks"

> or "Re-run failed checks" buttons*

Reviewer: ```suggestion
or "Re-run failed checks" buttons
```

- __triggers__: start of a new Precaution scan

> ## Events workflows

Reviewer: ```suggestion
## Event workflows
```

> This paragraph doesn't aim to explain in details how Precaution works.

Reviewer: ```suggestion
This paragraph doesn't aim to explain in detail how Precaution works.
```

# Events handling

## Subscribed events

Precaution needs to subscribe to a number of events

> to satisfy different use cases of our users.

Reviewer 2: "of our users" is redundant here.

The events are separated into two groups `pull requests
events` and `check suites/runs events`.

> ### Pull requests events

Reviewer 2: I thought we'd agreed to use the event name from the GitHub docs, PullRequestEvent, as the section title?

Official GitHub documentation: [PullRequest event](https://developer.github.com/v3/activity/events/types/#pullrequestevent)

* <strong>pull_request.opened</strong>

*occurs - when a new pull request is created*

*triggers - start of the Precaution scan*

> 

Reviewer 2: I'm still not a fan of the formatting choices here, what do you think of this?
#### `pull_request.opened`
- __occurs__: when a new pull request is created
- __triggers__: start of the Precaution scan

Author: I didn't know that markdown tricks.
A good idea.

* <strong>pull_request.reopened</strong>

*occurs - when a pull request is reopened from a closed state*

*triggers - start of a new Precaution scan*

* <strong>pull_request.synchronize</strong>

*occurs - when you add a commit to already opened pull request*

*triggers - start of a new Precaution scan*

> *for more information look at pull request number [#113](https://github.com/vmware/precaution/pull/113)*

Reviewer 2: ```suggestion
For more information look at pull request number [#113](https://github.com/vmware/precaution/pull/113)
```

Drop the italics and use sentence case

> ### Check run and check suite events

Reviewer 2: As above, didn't we agree to use theGitHub event names for the section titles?

Author: Yes, my mistake.

Official GitHub documentation:
- [CheckRunEvent](https://developer.github.com/v3/activity/events/types/#checkrunevent)
- [CheckSuiteEvent](https://developer.github.com/v3/activity/events/types/#checksuiteevent)

> A user had already opened a pull request and he is at the pull request page.

Reviewer 2: Drop redundant use of gendered pronoun
```suggestion
A user had already opened a pull request and is at the pull request page.
```

He navigates to the "Checks" tab.
If the Precaution check had failed he will see "Re-run all checks", "Re-run" and
"Re-run failed checks" buttons.

* <strong>check_run.rerequested</strong>

*occurs - when you click the "Re-run" button*

*triggers - start of a new Precaution scan*

* <strong>check_suite.rerequested</strong>

*occurs - when you click "Re-run all checks"
or "Re-run failed checks" buttons*

*triggers - start of a new Precaution scan*

## Events workflows

This paragraph doesn't aim to explain in details how Precaution works.

> If you want to read more details how Precaution works you can look at projects [Architecture](architecture.md).

Reviewer 2: This reads a bit awkwardly as there's repeated phrasing, I suggest we change the second sentence to something like

```suggestion
If you want to understand more details of Precaution's design see the [Architecture](architecture.md) documentation.
```

# Events handling

## Subscribed events

Precaution needs to subscribe to a number of events
to satisfy different use cases of our users.

The events are separated into two groups `pull requests
events` and `check suites/runs events`.

### Pull requests events

Official GitHub documentation: [PullRequest event](https://developer.github.com/v3/activity/events/types/#pullrequestevent)

* <strong>pull_request.opened</strong>

*occurs: when a new pull request is created*

*triggers: start of the Precaution scan*

* <strong>pull_request.reopened</strong>

*occurs - when a pull request is reopened from a closed state*

> *triggers: start of a new Precaution scan*

Reviewer 2: I feel like we need to delineate this triggers from `pull_request.opened`. "Re-start Precaution scan?"

Reviewer 2: Or maybe we just need to be consistent in all of the triggers. All triggered events are effectively the same, right? scanning all changes in the PR

Author: Yes, all triggers are the same. The pull request events are just different entry point into the Precaution workflow - which is basically a scan on all files in the pull request.

# Events handling

## Subscribed events

> Precaution needs to subscribe to a number of events when it's

Reviewer: "when it's to satisfy...." - sounds awkward.

to satisfy different use cases of our users.

> The events can be separated into two groups `pull requests

Reviewer: "can be" or "are"?

events` and `check suites/runs events`.

### Pull requests events

> Official GitHub documentation: [PullRequest event](https://developer.github.com/v3/activity/events/types/#pullrequestevent)

Reviewer: ```suggestion
Official GitHub documentation: [Pull Request event](https://developer.github.com/v3/activity/events/types/#pullrequestevent)
```

Reviewer 2: The event name in the GitHub docs is `PullRequestEvent`, I favour using that as the link title

Author: I agree with Joshua.

> * pull_request.opened - the first event which occurs when a new pull request is created. It's a self-explanatory

Reviewer: "It's a self-explanatory" - bad grammar.  Don't think this statement adds any value since you already explained it.

* pull_request.reopened - notifies Precaution to start a new scan on a reopened pull request

* pull_request.synchronize - occurs when you add a commit to already opened pull request.

> We discussed it and added it in pull request number [#113](https://github.com/vmware/precaution/pull/113)

Reviewer: Don't ever use "we" or "I" in documentation.  A reader doesn't know who "we" is.

> ### Check run/suite events

Reviewer: Why "/"  and not just "and"?

Author: Okay, maybe you are right. I will change it.

Official GitHub documentation:

> - [CheckRun event](https://developer.github.com/v3/activity/events/types/#checkrunevent)

Reviewer 2: ```suggestion
+ [CheckRunEvent](https://developer.github.com/v3/activity/events/types/#checkrunevent)
```

> - [CheckSuite event](https://developer.github.com/v3/activity/events/types/#checksuiteevent)

Reviewer 2: ```suggestion
- [CheckSuiteEvent](https://developer.github.com/v3/activity/events/types/#checksuiteevent)
```

> You had already opened a pull request and you are at the pull request page.

Reviewer: Avoid using "You"

Author: Maybe you are right. 
It will be better if I write from a third person view?

Then you navigate to the "Checks" tab.
If the Precaution check had failed you will see "Re-run all checks", "Re-run" and
"Re-run failed checks" buttons.

* check_run.rerequested - occurs when you click the "Re-run" button.
This will tell Precaution to start a new check run.

* check_suite.rerequested - happens when you click "Re-run all checks" 
or "Re-run failed checks" buttons. 

> This notifies Precaution and the other GitHub apps 

Reviewer: What other GitHub apps?  The context here is unclear.

Author: Okay, I won’t mention the other GitHub apps at all.

that they have to start a new check run.

## Events workflows

> This is a small paragraph about the events workflow we are receiving with and posting to GitHub.

Reviewer: Delete this line.  No point to describing what you will describe.

It doesn't aim to explain in details how Precaution works.

> If you want to read more details how Precaution works you can look at our [Architecture]().

Reviewer: Who is "our"???

Reviewer 2: the link to the architecture diagram is missing

There are numerous use cases but the triggered events and their order are alike.

> In the text below "F" equals "forked repository" and "U" means upstream repository.

Reviewer: Why use abbreviations?

Author: Because without them the titles of the different sections are too long. :D 
I will remove the abbreviations and you will tell me what you think.

Reviewer 2: You've stopped using the abbreviations below, so no need to define them.

### Pull request from F to U or from U to itself

1. `pull_request.opened` webhook received
2. `POST /repos/:owner/:repo/check-runs` - API call to create check_run
3. `GET /repos/:owner/:repo/contents/:.precaution.yaml` - API call to get `.precaution.yaml` config file
4. `GET /repos/:owner/:repo/pulls/:number/files` - API call to get list of modified files
5. `GET /repos/:owner/:repo/contents/:path` - multiple API calls to get the modified files
6. `PATCH /repos/:owner/:repo/check-runs/:check_run_id` - API call which sends the results

### Add commit to already opened pull request

1. `pull_request.synchronize webhook` webhook received
2. `POST /repos/:owner/:repo/check-runs` - API call to create check_run
3. `GET /repos/:owner/:repo/contents/:.precaution.yaml` - API call to get `.precaution.yaml` config file
4. `GET /repos/:owner/:repo/pulls/:number/files` - API call to get list of modified files
5. `GET /repos/:owner/:repo/contents/:path` - multiple API calls to get the modified files
6. `PATCH /repos/:owner/:repo/check-runs/:check_run_id` - API call which sends the results

> If you have already opened a pull request and you navigate to the "Checks" tab then

Reviewer: Run-on sentence

Reviewer 2: This duplicates information from the Check*Event section, no need to repeat it here.

Author: Yea, I will remove it. Probably we don’t need it.
